src/hash.cr

```diff
@@ -118,6 +290,479 @@ class Hash(K, V)
+        @indices = malloc_indices(indices_size)
+      else
+        return
+      end
+    else
+      # Only resize if we don't have many deleted elements.
+      if @deleted_count < @size
+        @entries = @entries.realloc(indices_size)
+        double_indices_size
+        @indices_bytesize = compute_indices_bytesize(indices_size)
+        @indices = realloc_indices(indices_size)
+      end
+
+      # Must must clear the indices because we'll rebuild them from scratch
+      clear_indices
+    end
+
+    # Here we traverse the `@entries` and compute their new index in `@indices`
+    # while moving non-deleted entries to the beginning (compaction).
+    new_entry_index = 0
+    each_entry_with_index do |entry, entry_index|
+      # First we move the index to its new index (if we need to do that)
+      set_entry(new_entry_index, entry) if entry_index != new_entry_index
+
+      # Then we try to find an empty index slot (we should find one now
+      # that we have more space)
+      index = fit_in_indices(entry.hash)
+      until get_index(index) == -1
+        index = next_index(index)
+      end
+
+      set_index(index, new_entry_index)
+      new_entry_index += 1
+    end
+
+    # We have to mark entries starting from the final new index
+    # as deleted so the GC can collect them.
+    entries_to_clear = entries_size - new_entry_index
+    if entries_to_clear > 0
+      (entries + new_entry_index).clear(entries_to_clear)
+    end
+
+    # After compaction we no longer have deleted entries
+    @deleted_count = 0
+  end
+
+  # After this it's 1 << 28, and with entries being Int32
+  # (4 bytes) it's 1 << 30 of actual bytesize and the
+  # next value would be 1 << 31 which overflows `Int32`.
+  private MAXIMUM_INDICES_SIZE = 1 << 28
+
+  # Doubles the value of `@indices_size` but first checks
+  # whether the maximum hash size is reached.
+  private def double_indices_size : Nil
+    if indices_size == MAXIMUM_INDICES_SIZE
+      raise "Maximim Hash size reached"
+    end
+
+    @indices_size_pow2 += 1
+  end
+
+  # Implementation of clearing the hash table.
+  private def clear_impl : Nil
+    @entries = Pointer(Entry(K, V)).null
+    @indices = Pointer(UInt8).null
+    @indices_size_pow2 = 0
+    @size = 0
+    @deleted_count = 0
+    @indices_bytesize = 1
+  end
+
+  # Gets from `@indices` at the given `index`.
+  # Returns the index in `@entries` or `-1` if the slot is empty.
+  private def get_index(index : Int32) : Int32
+    # Check what we have: UInt8, Int16 or UInt32 buckets
+    value = case @indices_bytesize
+            when 1
+              @indices[index].to_i32!
+            when 2
+              @indices.as(UInt16*)[index].to_i32!
+            else
+              @indices.as(UInt32*)[index].to_i32!
+            end
+    # Because 0 means empty we return -1,
+    # otherwise we decrement the value (so we can represent the index "0")
+    value == 0 ? -1 : value - 1
```

Comment:
value is Int32. Then why not just `value - 1` ?

Reply:
You are right! I don't know what I was thinking, hehe... thanks!